File: PTI.Rs232Validator/src/main/java/PTI/Rs232Validator/Loggers/Logger.java
package PTI.Rs232Validator.Loggers;


public abstract class Logger implements ILogger{


    protected Logger(String name, LogLevel minLogLevel) {
        Name = name;
        MinLogLevel = minLogLevel;
    }


    public String Name;


    public LogLevel MinLogLevel;


    public void LogTrace(String format, Object... args) {
        if(MinLogLevel.ordinal() > LogLevel.Trace.ordinal()){
            return;
        }

        Log(Name, LogLevel.Trace, format, args);
    }


    public void LogDebug(String format, Object... args) {
        if(MinLogLevel.ordinal() > LogLevel.Debug.ordinal()){
            return;
        }

        Log(Name, LogLevel.Debug, format, args);
    }


    public void LogInfo(String format, Object... args) {
        if(MinLogLevel.ordinal() > LogLevel.Info.ordinal()){
            return;
        }

        Log(Name, LogLevel.Info, format, args);
    }


    public void LogError(String format, Object... args) {
        if(MinLogLevel.ordinal() > LogLevel.Error.ordinal()){
            return;
        }

        Log(Name, LogLevel.Error, format, args);
    }


    protected abstract void Log(String Name, LogLevel level, String format, Object... args);
}

File: app/src/main/java/com/example/Rs232Validator/Logger.kt
package com.example.Rs232Validator

import PTI.Rs232Validator.Loggers.ILogger
import PTI.Rs232Validator.Loggers.LogLevel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

class Logger : ILogger {
    private val TimestampFormat: DateTimeFormatter = DateTimeFormatter.ofPattern("MM-dd-yyyy hh:mm:ss a")

    override fun LogTrace(format: String, vararg args: Any?) {
        Log(LogLevel.Trace, format, *args)
    }

    override fun LogDebug(format: String, vararg args: Any?) {
        Log(LogLevel.Debug, format, *args)
    }

    override fun LogInfo(format: String, vararg args: Any?) {
        Log(LogLevel.Info, format, *args)
    }

    override fun LogError(format: String, vararg args: Any?) {
        Log(LogLevel.Error, format, *args)
    }

    private fun Log(level: LogLevel, format: String, vararg args: Any?) {
        val entry = LogEntry(
            level,
            LocalDateTime.now().format(TimestampFormat),
            String.format(format, *args)
        )

        _LogEntries.update { it + entry }
    }

    private val _LogEntries = MutableStateFlow<List<LogEntry>>(emptyList())
    val logEntries: StateFlow<List<LogEntry>> = _LogEntries.asStateFlow()
}



data class LogEntry(
    val Level: LogLevel,
    val Timestamp: String,
    val Message: String
)